[Login] Improve error messaging displaying a new error screen when application password is disabled

pmusolino opened the pull request:
Closes: #14004 (and peaMlT-U4)

## Description
This pull request introduces improvements to handling the scenario when application passwords are disabled in user's website. The changes involve updating the error handling UI and navigation logic to provide a better user experience. This is a continuation of the job started by @hafizrahman [here](https://github.com/woocommerce/woocommerce-ios/pull/14269).

## Testing information
1. Create test site with WooCommerce and no Jetpack, on Jurassic Ninja.
2. Install and active a plugin like this one https://wordpress.org/plugins/disable-application-passwords/
3. Try to log in using application password to the test site.
4. Confirm that the new dialog is shown.
5. Confirm that tapping "Retry" button returns to two screens before the current one (instructions on how to approve the connection with application password).
6. Confirm that tapping "Log In With Another Account" returns to the initial screen of the login process.


## Screenshots

| Before             |  After |
:-------------------------:|:-------------------------:
![369337093-2987fa18-1589-4d07-9317-5cf5f7af9dac](https://github.com/user-attachments/assets/c16840a1-5e95-44c5-9cc4-a1178534408d) | <img src="https://github.com/user-attachments/assets/b470b1ab-3402-40cf-9b09-30d874316941" width=500 />


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

hafizrahman committed:
First pass at converting error dialog into an error screen for disabled application password.

pmusolino committed:
Merge branch 'trunk' into issue/14004-better-application-password-disabled-note

pmusolino committed:
Update navigation logic and button titles in ApplicationPasswordDisabledViewModel.swift
- Modify navigation logic to pop to the third last view controller or root if fewer than three
- Update primary button title to "Retry" for retrying application password authorization
- Remove outdated primary button title "Log in with WordPress.com"

pmusolino committed:
update: localization keys and message formatting in `ApplicationPasswordDisabledViewModel` using Woo mobile guidelines

pmusolino committed:
update: authentication tests and localization keys in `ApplicationPasswordViewModelTests`

pmusolino committed:
clean: removed TODO comment because the track already happen somewhere else

pmusolino committed:
update: release-notes 21.7

pmusolino committed:
update: added PR url in 21.7 release notes

itsmeichigo commented on WooCommerce/Classes/Authentication/Application Password/ApplicationPasswordAuthorizationWebViewController.swift:
```diff
@@ -171,6 +184,12 @@ private extension ApplicationPasswordAuthorizationWebViewController {
     }
 
+    @objc private func popBothControllers() {
+        // Pop back two view controllers to remove both error and web view
+        navigationController?.popViewController(animated: false)
+        navigationController?.popViewController(animated: true)
```
I think this is hacky and prone to error. How about looking for the index of `ApplicationPasswordTutorialViewController` in the navigation stack, and pop to the index before that? 

A better option would be to send the view controller before the application password flow to this view controller via `presentApplicationPasswordWebView` in `AuthenticationManager`.

pmusolino replied:
I agree with you and with your second suggestion, so I applied the new changes in this commit: 9b480ba48fc8c2f0c04aac77c43c512fb103c82e

itsmeichigo commented on WooCommerce/Classes/Authentication/Application Password/ApplicationPasswordAuthorizationWebViewController.swift:
```diff
@@ -254,6 +281,13 @@ extension ApplicationPasswordAuthorizationWebViewController: WKNavigationDelegat
     }
 }
 
+extension ApplicationPasswordAuthorizationWebViewController: UIAdaptivePresentationControllerDelegate {
+    func presentationControllerDidDismiss(_ presentationController: UIPresentationController) {
+        // This will be called when the error UI is dismissed
+        navigationController?.dismiss(animated: true)
+    }
+}
+
```
Is this still in use? I don't see where we present any modal.

pmusolino replied:
You're right, it seems unused, so I removed it 3f40c59fc1383b7881b655fb533c1c3aa81507dc

itsmeichigo commented on WooCommerce/Classes/Authentication/Navigation Exceptions/ApplicationPasswordDisabledViewModel.swift:
```diff
@@ -34,14 +37,18 @@ struct ApplicationPasswordDisabledViewModel: ULErrorViewModel {
-            return
+        guard let navigationController = viewController?.navigationController else { return }
+        let viewControllers = navigationController.viewControllers
+        if viewControllers.count >= 3 {
+            navigationController.popToViewController(viewControllers[viewControllers.count - 3], animated: true)
+        } else {
+            navigationController.popToRootViewController(animated: true)
         }
```
The number 3 seems to be a magic number, it's not clear why we have to do this. Please consider my suggestion above about looking for the index of a specific view controller to pop to.

pmusolino replied:
Now I'm using the `previousVC` like in the comment above, please feel free to check changes here: 9b480ba48fc8c2f0c04aac77c43c512fb103c82e 🙏

itsmeichigo reviewed (commented):
While testing navigation back and forth, I encountered a case where the application password fetch failed with the following log:

> Tracked application_password_authorization_url_fetch_failed, properties: [error_domain: Networking.RequestAuthenticatorError, error_description: Alamofire.AFError.requestAdaptationFailed(error: Networking.RequestAuthenticatorError.applicationPasswordNotAvailable), error_code: 1]

In this case, the `fetchAuthURL` method failed with error and the logic went through to the `catch` block, which showed an alert instead of the application password not available screen:


https://github.com/user-attachments/assets/f4418d46-e2b6-4c78-ba83-cf976dbe9d69

We should also check the logic in this catch block to show the new screen if we get the `applicationPasswordNotAvailable` error.

pmusolino committed:
Merge branch 'trunk' into issue/14004-better-application-password-disabled-note-take-2

pmusolino committed:
fix: remove unused code

pmusolino committed:
Add navigation to previous view controller in application password flow
- Modify `ApplicationPasswordAuthorizationWebViewController.swift` to include a reference to the previous view controller for navigation purposes.
- Update `AuthenticationManager.swift` to pass the previous view controller to the web view used for application password authorization.
- Enhance `ApplicationPasswordDisabledViewModel.swift` to store the previous view controller and use it for navigation.
- Adjust `PostSiteCredentialLoginChecker.swift` to capture and utilize the last view controller before the application password flow.
- Update tests in `ApplicationPasswordDisabledViewModelTests.swift` to include the previous view controller in the view model initialization.

pmusolino committed:
Merge branch 'trunk' into issue/14004-better-application-password-disabled-note-take-2

pmusolino commented:
@itsmeichigo this is ready for another check 🙌

itsmeichigo commented on WooCommerce/Classes/Authentication/Application Password/ApplicationPasswordAuthorizationWebViewController.swift:
```diff
@@ -34,6 +34,9 @@ final class ApplicationPasswordAuthorizationWebViewController: UIViewController
     }()
 
+    /// The view controller that was presenting the application password flow.
+    private weak var previousViewController: UIViewController?
```
Just wondering: is there any case where `previousViewController` can be deallocated when the user is still on `ApplicationPasswordAuthorizationWebViewController`? I'm curious why the weak reference is necessary here.

pmusolino replied:
I would say that it's more of a defensive approach. Do you think `previousViewController` might be deallocated if it's no longer needed or if the navigation stack changes (e.g., the user navigates away from the flow that involves `previousViewController`)?

itsmeichigo replied:
My understanding is that `previousViewController` should not be deallocated since it's part of the navigation stack that contains the `ApplicationPasswordAuthorizationWebViewController`.  The reason is that when one navigates back, the view controller and its states should still be there instead of being reinitialized. WDYT?

pmusolino replied:
I re-analyzed the whole flow, I think you are right, the whole flow will be deallocated all at once. I'm going to remove `weak`

itsmeichigo commented on WooCommerce/Classes/Authentication/Application Password/ApplicationPasswordAuthorizationWebViewController.swift:
```diff
@@ -171,6 +189,15 @@ private extension ApplicationPasswordAuthorizationWebViewController {
 
+    /// Pops to the previous view controller (if provided) or pops one level otherwise.
+    @objc private func navigateToPreviousViewController() {
+        if let previous = previousViewController, let nav = navigationController {
```
Super nit: we can simplify the `if let` checks now.

```suggestion
        if let previousViewController, let navigationController {
```

itsmeichigo commented on WooCommerce/Classes/Authentication/Navigation Exceptions/ApplicationPasswordDisabledViewModel.swift:
```diff
@@ -5,12 +5,20 @@ import WordPressAuthenticator
+    let image: UIImage = .errorImage
+
+    // The VC that was showing before the application password flow. This is used to navigate back without guesswork.
+    let previousViewController: UIViewController?
```
Related to a question above: the `previousViewController` reference is kept strongly here. Is it intentional to keep the weak reference in `ApplicationPasswordAuthorizationWebViewController` but strong here?

pmusolino replied:
No, it's not intentional. My initial idea was to use `weak` everywhere, but in this context, it's probably not needed. My doubt stems from not knowing all the possible navigation conditions in detail, so I opted to declare them as `weak` as a precaution. What do you think?

itsmeichigo replied:
As shared in the thread above, I think the previous view controller should not be deallocated while it's still in the navigation stack. It can only be released once you pop it from the stack.

itsmeichigo commented on WooCommerce/Classes/Authentication/Navigation Exceptions/ApplicationPasswordDisabledViewModel.swift:
```diff
@@ -34,14 +42,16 @@ struct ApplicationPasswordDisabledViewModel: ULErrorViewModel {
-        guard let viewController else {
-            return
+        guard let navigationController = viewController?.navigationController else { return }
+        if let previousViewController = previousViewController {
```
Super nit:

```suggestion
        if let previousViewController {
```

itsmeichigo commented on WooCommerce/Classes/Authentication/PostSiteCredentialLoginChecker.swift:
```diff
@@ -56,8 +56,9 @@ private extension PostSiteCredentialLoginChecker {
                 case ApplicationPasswordUseCaseError.applicationPasswordsDisabled:
-                    // show application password disabled error
-                    let errorUI = applicationPasswordDisabledUI(for: siteURL)
+                    // show application password disabled error, and catch the last view controller that was showing before the Application Password flow.
+                    let previousViewController = navigationController.viewControllers.dropLast().last
```
What is the exact controller that you are navigating to here? Please note that the `PostSiteCredentialLoginChecker` can be triggered from two places: (1) after site credential login (if application password is not required), and (2) after application password login. 

For (1), I'd suggest sending `nil` as the previous view controller to the `PostSiteCredentialLoginChecker`, so that the error page would simply pop itself upon navigating back.

For (2), I'd suggest sending the previous view controller from the application password entry point in `AuthenticationManager`

pmusolino replied:
Thank you for your suggestions @itsmeichigo, they are gold and I wasn't 100% aware of the flow. Can you check if the changes make sense now, please? 6bdb62db7e1266caa2dd63dc920b502d2ad83f4b

itsmeichigo commented on WooCommerce/Classes/Authentication/Navigation Exceptions/ApplicationPasswordDisabledViewModel.swift:
```diff
@@ -34,14 +42,16 @@ struct ApplicationPasswordDisabledViewModel: ULErrorViewModel {
+        if let previousViewController = previousViewController {
+            navigationController.popToViewController(previousViewController, animated: true)
+        } else {
+            navigationController.popToRootViewController(animated: true)
```
Since you want to pop to the previous view controller, `popToRoot` seems incorrect. Should it be `popViewController`?

pmusolino replied:
When I implemented it, during my testing process (at least in every test I did), the previous view controller immediately before the error screen was the loading screen—the one with just a loader in the middle of the view. If we pop back to it (by calling `popViewController`), the user would remain stuck there, so we need to go back further to ensure a better user experience.

itsmeichigo replied:
Thanks for the explanation. From my understanding, you already sent the `previousViewController` for the application password flow, so this should not happen. Also, you are sending `nil` for the site credential login flow, so tapping "Try Again" should navigate back to the login screen instead of the prologue screen IMO. Is there any other edge case we should be aware of?

pmusolino replied:
Yes, in theory, there are no other edge cases and I didn't find any other specific edge cases so I'm putting `popViewController`. 👍

itsmeichigo reviewed (changes requested):
Thanks for the updates, Paolo. I found a few issues with the new changes in the comments below. 

Also, it seems that you missed the comment in my previous round about missing the case when fetching the authentication URL fails with `Alamofire.AFError.requestAdaptationFailed(error: Networking.RequestAuthenticatorError.applicationPasswordNotAvailable)`.

pmusolino committed:
Update WooCommerce/Classes/Authentication/Application Password/ApplicationPasswordAuthorizationWebViewController.swift

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

pmusolino committed:
Update WooCommerce/Classes/Authentication/Navigation Exceptions/ApplicationPasswordDisabledViewModel.swift

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

pmusolino committed:
fix: naming

pmusolino committed:
Add previousViewController parameter for navigation
- Update `AuthenticationManager.swift` to include `previousViewController` parameter in `checkSiteCredentialLogin` function
- Modify `PostSiteCredentialLoginChecker.swift` to accept `previousViewController` in the initializer and utilize it in error handling

pmusolino committed:
Add previousViewController parameter to PostSiteCredentialLoginCheckerTests: update `PostSiteCredentialLoginChecker` instantiation in `PostSiteCredentialLoginCheckerTests.swift` to include `previousViewController: nil` across multiple test cases.

pmusolino commented:
> Also, it seems that you missed the comment in my previous round about missing the case when fetching the authentication URL fails with Alamofire.AFError.requestAdaptationFailed(error: Networking.RequestAuthenticatorError.applicationPasswordNotAvailable).

I'm sorry for missing that. How is it possible to generate that specific error? Is this something you simulate by editing the code, or is there a specific case that you can consistently replicate on JN sites? Note that simply disabling application passwords doesn't generate that error for me on my side.

itsmeichigo commented:
> I'm sorry for missing that. How is it possible to generate that specific error? Is this something you simulate by editing the code, or is there a specific case that you can consistently replicate on JN sites? Note that simply disabling application passwords doesn't generate that error for me on my side.

As stated above, I encountered this by navigating back and forth from the application password disabled error screen. Specifically:
- Follow the steps to the application password login flow until the disabled error screen.
- Tap Retry or "<" to retry login.
- If the same disabled screen is display, try again.
- After a few times, I got the error `Alamofire.AFError.requestAdaptationFailed(error: Networking.RequestAuthenticatorError.applicationPasswordNotAvailable)` for the URL fetching request, and an alert is displayed as the video above.

Though this sounds like an edge case, I believe it should be handled given 2 reasons:
- The application password flow is unpredictable for different sites.
- It makes sense to handle the same issue when the error clearly is `applicationPasswordNotAvailable`.

pmusolino committed:
Merge branch 'trunk' into issue/14004-better-application-password-disabled-note-take-2

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/237">21.7</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

pmusolino committed:
feat: remove weak from `ApplicationPasswordAuthorizationWebViewController`

pmusolino committed:
update: navigation logic in `ApplicationPasswordDisabledViewModel`: change navigationController method from `popToRootViewController` to `popViewController`

pmusolino commented:
@itsmeichigo I was able to replicate the issue you mentioned when going back and forth. Do you think the error should be handled with the same error screen, or should we show a different error? I'm not sure if the error is always due to the application password being disabled or something else, because if I enable the application password, replicating that specific error becomes impossible, at least in my tests. If we display an error regarding fetching the URL, it will be generic and not helpful for the user, so I think I can move forward by displaying the same error screen about the application password being disabled. What do you think?

pmusolino committed:
Merge branch 'trunk' into issue/14004-better-application-password-disabled-note-take-2

itsmeichigo commented:
@pmusolino I debugged the app and found that the flow now doesn't go through web application password, so we've been testing the wrong thing. Please note: you can only reach the web application password flow if you see the tutorial page before opening the web view. 

I used Proxyman to check the requests and here are what I'm seeing with a JN site that has https://wordpress.org/plugins/disable-application-passwords/ installed, and this is the flow I see up until the error reported above:

<img width="1030" alt="image" src="https://github.com/user-attachments/assets/f228bec9-b052-4a46-9d58-9044d80f303a" />

So the flow first went through the application password retrieval with nonce (handled by `DefaultApplicationPasswordUseCase` in the app, then failed when the POST request `rest_route=/wp/v2/users/me/application-passwords` failed. The application password disabled screen displayed here was added by Hafiz, not the implementation added in this PR.

When I navigated back and forth many times to retry, I finally got the 429 error (too many requests) for the nonce-retrieval request (`wp-admin/admin-ajax.php?action=rest-nonce`). Only then did I got redirected to the web application password flow and saw the error reported above.

---

Now, to actually reach the application password web view screen, we need to block the programmatic site credential login through `wp-login.php`. There are a few ways to do this: customize the login URL for the wp-admin page, enable captcha on the test site, etc. I enabled captcha for my site's login page by adding WordFence plugin to my site and enabling reCAPTCHA to block auto-login.

 As can be seen in the screencast below, after entering site credentials, I was redirected to the web application password tutorial page with the error at the top. Then when I proceeded, I immediately got the error alert for disabled application password. This means it's a must to display the disabled error screen upon catching error when fetching application password URL.


https://github.com/user-attachments/assets/bb2b3b88-fba5-4982-b06c-915a7fbfff67



---

Regarding your question:

> I'm not sure if the error is always due to the application password being disabled or something else, because if I enable the application password, replicating that specific error becomes impossible, at least in my tests

A few suggestions in case it's helpful:
- Please ensure to understand the application password flow thoroughly by taking a closer look at the implementation.
- Please be more proactive in debugging when you see something you cannot explain.

pmusolino committed:
Refactor error handling in `ApplicationPasswordAuthorizationWebViewController` and display errorUI instead of error alert

pmusolino commented:
Thank you so much Huong for the explanation.
I did another test and was able to replicate it, and I saw your same [requests](https://github.com/user-attachments/assets/2c9da253-b63c-4719-8468-7e456bf35a02). I replaced the alert with the new UI error screen and tomorrow I'll check If I need to do other changes. In the meantime can you take a look?9fafc530cd0f85712d6e925be4c41ab80cca27a2

**Extra:**
While testing, I gained insights into how we're currently managing error codes related to application passwords. I've identified a scenario that may not be easily addressed and might have been discussed previously. In a new JN site, before encountering any 429 errors and with application passwords enabled, installing a plugin such as `WPS Hide Login`—which relocates the login page—causes the webpage to load but displays a "Page not found" message instead of an error (see image below).

<img src="https://github.com/user-attachments/assets/080d9dbc-10aa-4d2f-a597-bdb5de12a9cc" width=300 />

I believe this happens because the `handleSiteCredentialLoginFailure` function in the `AuthenticationManager` returns a generic `SiteCredentialLoginError`. Although we can recognize this situation as `inaccessibleLoginPage` under the 404 error code, and we include a tutorial message stating, "This could be because your store has some extra security steps in place," we still direct the user to the webpage. This occurs because we can't distinguish this specific error from others that also return a 404 code. From a UI/UX perspective, this isn't ideal for the end user, and it remains a challenging issue to identify and resolve, especially if we have no other information on the page with which to differentiate an authentication page from an error page.

pmusolino committed:
Merge branch 'trunk' into issue/14004-better-application-password-disabled-note-take-2

itsmeichigo commented:
> In a new JN site, before encountering any 429 errors and with application passwords enabled, installing a plugin such as WPS Hide Login—which relocates the login page—causes the webpage to load but displays a "Page not found" message instead of an error

Regarding this, I believe `WPS Hide Login` has the option to redirect to 404 immediately when accessing the wp-admin directory. The authentication URL belongs to this directory by default, hence the Page not found that you saw.

<img width="1001" alt="image" src="https://github.com/user-attachments/assets/a8bc049d-bc5f-4272-b2ba-3360e7c0966d" />

So, this is not an issue with the customization of the login page, but a custom rule of redirect to 404. In this case, I don't think we can do anything, users will need to contact support for help. cc @hichamboushaba in case he has some other suggestions.

itsmeichigo commented:
> In a new JN site, before encountering any 429 errors and with application passwords enabled, installing a plugin such as WPS Hide Login—which relocates the login page—causes the webpage to load but displays a "Page not found" message instead of an error

Regarding this, I believe `WPS Hide Login` has the option to redirect to 404 immediately when accessing the wp-admin directory. The authentication URL belongs to this directory by default, hence the Page not found that you saw.

<img width="1001" alt="image" src="https://github.com/user-attachments/assets/a8bc049d-bc5f-4272-b2ba-3360e7c0966d" />

So, this is not an issue with the customization of the login page, but a custom rule of redirect to 404. In this case, I don't think we can do anything, users will need to contact support for help. cc @hichamboushaba in case he has some other suggestions.

pmusolino committed:
Merge branch 'trunk' into issue/14004-better-application-password-disabled-note-take-2

pmusolino commented:
> Regarding this, I believe WPS Hide Login has the option to redirect to 404 immediately when accessing the wp-admin directory. The authentication URL belongs to this directory by default, hence the Page not found that you saw.

Yes, I think that it's a very common behavior for this type of plugin.

itsmeichigo commented on RELEASE-NOTES.txt:
```diff
@@ -1,8 +1,13 @@
 21.7
 -----
 - [**] Fixed an issue with the WordPress Media Library not loading due to a decoding error in media details sizes. [https://github.com/woocommerce/woocommerce-ios/pull/15056]
+- [*] Better error messages if Application Password login is disabled on user's website. [https://github.com/woocommerce/woocommerce-ios/pull/15031]
```
This should now be moved to 21.8.

itsmeichigo commented on WooCommerce/Classes/Authentication/Application Password/ApplicationPasswordAuthorizationWebViewController.swift:
```diff
@@ -157,20 +162,41 @@ private extension ApplicationPasswordAuthorizationWebViewController {
+                    return
                 }
                 loadAuthorizationPage(url: url)
             } catch {
                 DDLogError("⛔️ Error fetching authorization URL for application passwords \(error)")
                 analytics.track(.applicationPasswordAuthorizationURLFetchFailed, withError: error)
-                showErrorAlert(message: Localization.errorFetchingAuthURL, onRetry: { [weak self] in
-                    self?.fetchAuthorizationURL()
-                })
+                navigateToErrorUI()
```
To be safe, we should rename `navigateToErrorUI` to `navigateToApplicationPasswordDisabledUI` and trigger it only when catching `Alamofire.AFError.requestAdaptationFailed(error: Networking.RequestAuthenticatorError.applicationPasswordNotAvailable)`

pmusolino replied:
I think then we should maintain the error alert for all other cases. This way, we can address any unexpected error that might arise other then the application password not being available

itsmeichigo commented on WooCommerce/Classes/Authentication/Application Password/ApplicationPasswordAuthorizationViewModel.swift:
```diff
@@ -4,7 +4,7 @@ import Yosemite
 ///
 final class ApplicationPasswordAuthorizationViewModel {
-    private let siteURL: String
+    let siteURL: String
```
Seems like this change is not necessary?

pmusolino replied:
The function `applicationPasswordDisabledUI` under `navigateToErrorUI` uses the `siteURL`, so it needs to be declared as internal now.

pmusolino committed:
Merge branch 'trunk' into issue/14004-better-application-password-disabled-note-take-2

pmusolino committed:
update: moved release notes to v21.8

pmusolino committed:
- Change `RequestAuthenticatorError` enum to public in `RequestAuthenticator.swift`
- Import `RequestAuthenticatorError` in `ApplicationPasswordAuthorizationWebViewController.swift`
- Update error navigation method name to `navigateToApplicationPasswordDisabledUI` in `ApplicationPasswordAuthorizationWebViewController.swift`
- Modify error handling logic to differentiate between `applicationPasswordNotAvailable` and other errors in `ApplicationPasswordAuthorizationWebViewController.swift`

itsmeichigo commented on RELEASE-NOTES.txt:
```diff
@@ -11,10 +11,14 @@
 21.7
 -----
 - [**] Fixed an issue with the WordPress Media Library not loading due to a decoding error in media details sizes. [https://github.com/woocommerce/woocommerce-ios/pull/15056]
+- [*] Now, usernames and emails in text fields across multiple login views are no longer capitalized. [https://github.com/woocommerce/woocommerce-ios/pull/15002]
+- [*] Product Details: Display cover tag on the first product image [https://github.com/woocommerce/woocommerce-ios/pull/15041]
+- [*] Payments: Update learn more links to open Stripe-specific docs when using that gateway [https://github.com/woocommerce/woocommerce-ios/pull/15035]
```
These seem to be duplicated and should be removed.

itsmeichigo commented on WooCommerce/Classes/Authentication/Application Password/ApplicationPasswordAuthorizationWebViewController.swift:
```diff
@@ -157,20 +163,46 @@ private extension ApplicationPasswordAuthorizationWebViewController {
+                   authError == .applicationPasswordNotAvailable {
+                    navigateToApplicationPasswordDisabledUI()
+                } else {
+                    showErrorAlert(message: Localization.authorizationRejected)
```
We should display `errorFetchingAuthURL` instead of `authorizationRejected` here. You can restore `errorFetchingAuthURL` from the git history.

pmusolino replied:
👀 5dd3e8349accbf0df2ff0b5259705a7faec22027 Thanks for noticing it!

pmusolino committed:
Merge branch 'trunk' into issue/14004-better-application-password-disabled-note-take-2

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/238">21.8</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

pmusolino committed:
update: change error alert message in `showErrorAlert` from `Localization.authorizationRejected` to `Localization.errorFetchingAuthURL`
